File: light_delivery/light_delivery/report/delivery_valuation/delivery_valuation.py
# Copyright (c) 2024, Muhammad Essam and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def get_delivery_report(filters=None):
    if not filters:
        filters = {}

    deliveries = frappe.get_all('Delivery',
        fields=['name', 'delivery_name', 'date_of_joining', 'delivery_category', 'valuation'],
        filters=filters
    )

    report_data = []
    
    for delivery in deliveries:
        delivery_name = delivery['name']
        
        request_data = frappe.db.sql("""
            SELECT
                SUM(CASE WHEN d_r.status = 'Delivered' THEN 1 ELSE 0 END) AS total_delivered,
                SUM(CASE WHEN d_r.status IN ('Delivery Cancel', 'Store Cancel', 'Cancel') THEN 1 ELSE 0 END) AS total_canceled
            FROM `tabRequest Delivery` as d_r
            WHERE delivery = %s
        """, delivery_name, as_dict=True)
        
        if request_data:
            delivery.update({
                'total_delivered': request_data[0].total_delivered,
                'total_canceled': request_data[0].total_canceled
            })

        report_data.append(delivery)

    return report_data

def get_report_columns():
    return [
        {"label": _("Delivery Name"), "fieldname": "name", "fieldtype": "Link", "options": "Delivery", "width": 180},
		{"label": _("Delivery Name"), "fieldname": "delivery_name", "fieldtype": "Data", "width": 180},
        {"label": _("Date of Joining"), "fieldname": "date_of_joining", "fieldtype": "Date", "width": 150},
        {"label": _("Delivery Category"), "fieldname": "delivery_category", "fieldtype": "Link", "options": "Delivery Category", "width": 180},
        {"label": _("Valuation"), "fieldname": "valuation", "fieldtype": "Rating", "width": 150},
        {"label": _("Total Requested Delivered"), "fieldname": "total_delivered", "fieldtype": "Int", "width": 150},
        {"label": _("Total Requested Canceled"), "fieldname": "total_canceled", "fieldtype": "Int", "width": 150}
    ]


def execute(filters=None):
    columns = get_report_columns()
    data = get_delivery_report(filters)
    return columns, data

File: light_delivery/light_delivery/report/delivery_valuation/delivery_valuation.js
// Copyright (c) 2024, Muhammad Essam and contributors
// For license information, please see license.txt

frappe.query_reports["Delivery Valuation"] = {
	"filters": [
        {
            "fieldname": "delivery_category",
            "label": __("Delivery Category"),
            "fieldtype": "Link",
            "options": "Delivery Category",
            "default": "",
            "reqd": 0
        },
        {
            "fieldname": "date_of_joining",
            "label": __("Date of Joining"),
            "fieldtype": "Date",
            "default": "",
            "reqd": 0
        },
		{
            "fieldname": "valuation",
            "label": __("Valuation"),
            "fieldtype": "Rating",
            "default": "",
            "reqd": 0
        },
    ],

	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		// Center the text in specific columns
		if (column.fieldname === "name" || 
			column.fieldname === "delivery_name" || 
			column.fieldname === "date_of_joining" || 
			column.fieldname === "delivery_category" || 
			column.fieldname === "valuation" ||
			column.fieldname === "total_delivered" || 
			column.fieldname === "total_canceled") {
			
			value = `<div style="text-align: center;">${value}</div>`;
		}

		return value;
	}

}
